pkg/validator/cluster.go

```diff
@@ -312,13 +312,79 @@ func validateGeneralClusterChanges(
 	valErrors []string,
 ) []string {
 
-	if cr.Spec.AppID != prevCr.Spec.AppID {
-		appModifiedMsg := fmt.Sprintf(
-			modifiedProperty,
-			"app",
+	readyToUpgrade := shared.ClusterIsReady(prevCr)
+	upgradeInfo := prevCr.Status.UpgradeInfo
+	// If changed cluster has actual upgradeInfo (was in Updating state)
+	// we may permit the changes for rollback case only
+	// It means, the current proposed AppId should be the same as the stored in the upgradeInfo object
+	if upgradeInfo != nil && !upgradeInfo.IsRollingBack {
+		readyToUpgrade = cr.Spec.AppID == upgradeInfo.PrevApp
+	}
+
+	if !readyToUpgrade {
+		clusterNotReadyMsg := fmt.Sprintf(
+			clusterNotReady,
+			prevCr.Name,
+			prevCr.Status.State,
 		)
-		valErrors = append(valErrors, appModifiedMsg)
+		valErrors = append(valErrors, clusterNotReadyMsg)
+		return valErrors
+	}
+
+	crApp, err := catalog.GetApp(cr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
 	}
+
+	prevCrApp, err := catalog.GetApp(prevCr)
+	if err != nil {
+		valErrors = append(valErrors, err.Error())
+	}
+
+	appModified := cr.Spec.AppID != prevCr.Spec.AppID
+
+	if appModified {
+		// Spec change not allowed if a user tries to update spec.app with other spec fields in the same time.
+		prevAppId := prevCr.Spec.AppID
+		prevCr.Spec.AppID = cr.Spec.AppID
+
+		if !equality.Semantic.DeepEqual(cr.Spec, prevCr.Spec) {
+			valErrors = append(valErrors, notOnlyAppModified)
+			return valErrors
+		}
+		prevCr.Spec.AppID = prevAppId
+
+		// DistroID should stay the same
+		if prevCrApp.Spec.DistroID != crApp.Spec.DistroID {
+			appModifiedMsg := fmt.Sprintf(
+				invalidDistroId,
+				crApp.Spec.DistroID,
+				prevCrApp.Spec.DistroID,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		// App version should be different from previous one
+		if prevCrApp.Spec.Version == crApp.Spec.Version {
+			appModifiedMsg := fmt.Sprintf(
+				versionIsNotModified,
+				crApp.Spec.DistroID,
+				crApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+
+		// Also, the current app should support upgrade
+		if !prevCrApp.Spec.Upgradable {
+			appModifiedMsg := fmt.Sprintf(
+				appNotUpgradable,
+				prevCrApp.Spec.DistroID,
+				prevCrApp.Spec.Version,
+			)
+			valErrors = append(valErrors, appModifiedMsg)
+		}
+	}
```

Comment:
It looks like we still need the following check:

While an upgrade is in progress (i.e. was initiated on some previous handler, so upgradeInfo exists), you're not allowed to make any spec changes other than rollback.

Reply:
Isn't this check not enough?

```
	if appModified {
		// Spec change not allowed if a user tries to update spec.app with other spec fields in the same time.
		prevAppId := prevCr.Spec.AppID
		prevCr.Spec.AppID = cr.Spec.AppID

		if !equality.Semantic.DeepEqual(cr.Spec, prevCr.Spec) {
			valErrors = append(valErrors, notOnlyAppModified)
			return valErrors
		}
		prevCr.Spec.AppID = prevAppId
```

Reply:
Nope because appModified will not be true on subsequent reconciliation handlers during the upgrade.

Reply:
Well, I adjusted this logic